File: experiments/test_multi_reference.js
/**
 * Multi-Reference Feature Experiment (Issue #184)
 *
 * This script tests the concept of multi-references where
 * multiple space-separated words before a colon form a single reference.
 */

import { Parser, Link, formatLinks } from '../js/src/index.js';

const parser = new Parser();

console.log('=== Multi-Reference Feature Tests (Issue #184) ===\n');

// Test 1: Single-word ID (backward compatibility)
const test1 = 'papa: loves mama';
console.log('Test 1 - Single-word ID (backward compatible):');
console.log('Input:', test1);
try {
  const result1 = parser.parse(test1);
  console.log('Parsed:', JSON.stringify(result1, null, 2));
  console.log('Formatted:', formatLinks(result1, true));
  console.log('✅ Pass: Single-word ID still works');
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

// Test 2: Quoted multi-word ID (backward compatibility)
const test2 = "('some example': value)";
console.log('Test 2 - Quoted multi-word ID (backward compatible):');
console.log('Input:', test2);
try {
  const result2 = parser.parse(test2);
  console.log('Parsed:', JSON.stringify(result2, null, 2));
  console.log('Formatted:', formatLinks(result2, true));
  console.log('✅ Pass: Quoted multi-word ID still works');
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

// Test 3: Unquoted multi-word ID (NEW FEATURE)
const test3 = '(some example: some example is a link)';
console.log('Test 3 - Unquoted multi-word ID (NEW):');
console.log('Input:', test3);
try {
  const result3 = parser.parse(test3);
  console.log('Parsed:', JSON.stringify(result3, null, 2));
  console.log('Formatted:', formatLinks(result3, true));
  // Check if ID is an array with 2 elements
  if (Array.isArray(result3[0].id) && result3[0].id.length === 2) {
    console.log('✅ Pass: Multi-reference ID parsed as array:', result3[0].id);
  } else {
    console.log('⚠️ ID is not an array:', result3[0].id);
  }
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

// Test 4: Context-aware multi-reference recognition in values
const test4 = '(some example: some example is a link)';
console.log('Test 4 - Context-aware multi-reference in values:');
console.log('Input:', test4);
try {
  const result4 = parser.parse(test4);
  console.log('Values count:', result4[0].values.length);
  console.log('First value:', result4[0].values[0]);
  // Check if "some example" in values is recognized as a single multi-ref
  if (Array.isArray(result4[0].values[0].id) &&
      result4[0].values[0].id.length === 2 &&
      result4[0].values[0].id[0] === 'some' &&
      result4[0].values[0].id[1] === 'example') {
    console.log('✅ Pass: "some example" recognized as multi-reference in values');
  } else {
    console.log('⚠️ Multi-reference not recognized:', result4[0].values[0].id);
  }
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

// Test 5: Multiple multi-references in one document
const test5 = `(some example: some example is a link)
some example`;
console.log('Test 5 - Self-reference (multi-ref used standalone):');
console.log('Input:', test5);
try {
  const result5 = parser.parse(test5);
  console.log('Parsed links count:', result5.length);
  console.log('Second link:', JSON.stringify(result5[1], null, 2));
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

// Test 6: Mixed references (single and multi)
const test6 = '(new york city: new york city is great)';
console.log('Test 6 - Three-word multi-reference:');
console.log('Input:', test6);
try {
  const result6 = parser.parse(test6);
  console.log('Parsed:', JSON.stringify(result6, null, 2));
  console.log('ID:', result6[0].id);
  console.log('Values count:', result6[0].values.length);
  if (Array.isArray(result6[0].id) && result6[0].id.length === 3) {
    console.log('✅ Pass: 3-word multi-reference parsed correctly');
  }
} catch (e) {
  console.log('❌ Fail:', e.message);
}
console.log();

console.log('=== Summary ===\n');
console.log('Multi-reference feature implemented:');
console.log('1. Grammar updated to allow multiple references before colon');
console.log('2. ID field can now be string (single) or string[] (multi)');
console.log('3. Context-aware recognition: defined multi-refs recognized in values');
console.log('4. Backward compatible: single-word and quoted IDs still work');

File: experiments/test_multi_reference.py
#!/usr/bin/env python3
"""
Multi-Reference Feature Experiment (Issue #184)

This script tests the concept of multi-references where
multiple space-separated words before a colon form a single reference.
"""

import sys
sys.path.insert(0, 'python')

from links_notation import Parser, format_links

parser = Parser()

print("=== Multi-Reference Feature Tests (Issue #184) - Python ===\n")

# Test 1: Single-word ID (backward compatibility)
test1 = "papa: loves mama"
print("Test 1 - Single-word ID (backward compatible):")
print("Input:", test1)
try:
    result1 = parser.parse(test1)
    print("Parsed ID:", result1[0].id)
    print("Values:", [v.id for v in result1[0].values])
    print("Formatted:", format_links(result1, True))
    if isinstance(result1[0].id, str):
        print("✅ Pass: Single-word ID still works as string")
    else:
        print("⚠️ Warning: ID type changed")
except Exception as e:
    print("❌ Fail:", e)
print()

# Test 2: Quoted multi-word ID (backward compatibility)
test2 = "('some example': value)"
print("Test 2 - Quoted multi-word ID (backward compatible):")
print("Input:", test2)
try:
    result2 = parser.parse(test2)
    print("Parsed ID:", result2[0].id)
    print("Formatted:", format_links(result2, True))
    if isinstance(result2[0].id, str) and result2[0].id == "some example":
        print("✅ Pass: Quoted multi-word ID still works as string")
    else:
        print("⚠️ Warning: ID type changed")
except Exception as e:
    print("❌ Fail:", e)
print()

# Test 3: Unquoted multi-word ID (NEW FEATURE)
test3 = "(some example: some example is a link)"
print("Test 3 - Unquoted multi-word ID (NEW):")
print("Input:", test3)
try:
    result3 = parser.parse(test3)
    print("Parsed ID:", result3[0].id)
    print("Values:", [v.id for v in result3[0].values])
    print("Formatted:", format_links(result3, True))
    if isinstance(result3[0].id, list) and result3[0].id == ["some", "example"]:
        print("✅ Pass: Multi-reference ID parsed as list:", result3[0].id)
    else:
        print("⚠️ ID is not a list:", result3[0].id)
except Exception as e:
    print("❌ Fail:", e)
print()

# Test 4: Context-aware multi-reference recognition in values
test4 = "(some example: some example is a link)"
print("Test 4 - Context-aware multi-reference in values:")
print("Input:", test4)
try:
    result4 = parser.parse(test4)
    print("Values count:", len(result4[0].values))
    print("First value ID:", result4[0].values[0].id)
    # Check if "some example" in values is recognized as a single multi-ref
    if (isinstance(result4[0].values[0].id, list) and
        result4[0].values[0].id == ["some", "example"]):
        print("✅ Pass: 'some example' recognized as multi-reference in values")
    else:
        print("⚠️ Multi-reference not recognized:", result4[0].values[0].id)
except Exception as e:
    print("❌ Fail:", e)
print()

# Test 5: Three-word multi-reference
test5 = "(new york city: new york city is great)"
print("Test 5 - Three-word multi-reference:")
print("Input:", test5)
try:
    result5 = parser.parse(test5)
    print("Parsed ID:", result5[0].id)
    print("Values count:", len(result5[0].values))
    if isinstance(result5[0].id, list) and len(result5[0].id) == 3:
        print("✅ Pass: 3-word multi-reference parsed correctly")
    else:
        print("⚠️ Unexpected result")
except Exception as e:
    print("❌ Fail:", e)
print()

# Test 6: Indented syntax with multi-reference
test6 = """some example:
  value1
  value2"""
print("Test 6 - Indented syntax with multi-reference:")
print("Input:", repr(test6))
try:
    result6 = parser.parse(test6)
    print("Parsed ID:", result6[0].id)
    print("Values count:", len(result6[0].values))
    if isinstance(result6[0].id, list) and result6[0].id == ["some", "example"]:
        print("✅ Pass: Indented multi-reference works")
    else:
        print("⚠️ Unexpected result")
except Exception as e:
    print("❌ Fail:", e)
print()

print("=== Summary ===\n")
print("Multi-reference feature implemented in Python:")
print("1. Parser updated to support multi-word IDs before colon")
print("2. ID field can now be string (single) or list[str] (multi)")
print("3. Context-aware recognition: defined multi-refs recognized in values")
print("4. Backward compatible: single-word and quoted IDs still work")
